### ISSUE TITLE
[jsx-closing-tag-location]: Add line-aligned option similar to jsx-closing-bracket-location

### ISSUE THREAD

**Issue by NiGhTTraX** [yannickcr/eslint-plugin-react#1466]
## Current behavior

This warns:

```jsx
return <div>
  foobar
</div>;
```

This doesn't:

```jsx
return (
  <div>
    foobar
  </div>
);
```

The rule is checking that the column of the closing tag is equal to the column of the opening tag:

https://github.com/yannickcr/eslint-plugin-react/blob/a120758e25f04cdc37d2d3d80922f2b998ab63ed/lib/rules/jsx-closing-tag-location.js#L52-L59

## Expected behavior

Ideally, this rule would accept an option `line-aligned` like [jsx-closing-bracket-location](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-closing-bracket-location.md#location) that would enforce the closing tag to be aligned with the line containing the opening line.

I'm assuming the code would be similar to:

https://github.com/yannickcr/eslint-plugin-react/blob/a120758e25f04cdc37d2d3d80922f2b998ab63ed/lib/rules/jsx-closing-bracket-location.js#L115-L126

**Comment by ljharb** [comment]
I think I understand what you're asking for, but not "why". Certainly this non-JSX is valid:
```js
return [
  foo,
];
```

but in that case, the array is part of javascript. with jsx, an opening tag enters you into an _entirely distinct parsing context_, and it's incredibly important to denote that by giving it a fresh indentation context.

Since we have "props-aligned", "line-aligned" might indeed make sense to add, but I really don't understand the benefit of this style.

**Comment by NiGhTTraX** [comment]
Sorry @ljharb, I'm not sure I understand your example. Maybe I wasn't clear enough in my issue. I want the following to be valid indentation:

```jsx
render() {
  return <div>
    foobar
  </div>;
}
```

The rule wants this instead:

```jsx
render() {
  return <div>
    foobar
         </div>;
}
```

**Comment by ljharb** [comment]
Right; your first example is what I'm suggesting isn't a good practice - but I do understand that's what you want.

Your second example, however, seems like a clear bug. If the setting is "tag-aligned", then "foobar" should be indented like so:
```jsx
render() {
  return <div>
           foobar
         </div>;
}
```
or like this:
```jsx
render() {
  return (
    <div>
      foobar
    </div>
  );
}
```

"line-aligned" would then be like this:
```jsx
render() {
  return <div>
    foobar
  </div>;
}
```

I think it makes sense for closing-tag-location to have the same options as closing-bracket-location, despite my personal opinion on "line-aligned".

**Comment by eddiemonge** [comment]
Maybe you should just do JSX like this rule instead? https://github.com/yannickcr/eslint-plugin-react/blob/HEAD/docs/rules/jsx-wrap-multilines.md

**Comment by brammitch** [comment]
Another use case for "line-aligned":

We use the [Operator-Linebreak with "Before" option](https://eslint.org/docs/rules/operator-linebreak#before) to make ternary statements easier to read.

Without the "line-aligned" option, this generates a warning:
```
{this.props.isLoading
  ? <ProgressBar>
  : <div>
    foo
  </div>
}
```

**Comment by ljharb** [comment]
@brammitch
```jsx
{this.props.isLoading
  ? <ProgressBar>
  : (
    <div>
      foo
    </div>
  )
}
```
or
```jsx
{this.props.isLoading ? <ProgressBar> : (
  <div>
    foo
  </div>
)}
```

but if you don't wrap multiline jsx in parens for some reason, then i can see your use case applying.

**Comment by brammitch** [comment]
@ljharb 
We don't wrap multiline JSX in parens.

Looking at Airbnb's style guide, it seems like we should change our rules to use jsx-wrap-multilines instead of making an exception for line-aligned closing tags.

**Comment by mrchief** [comment]
Here's a use case:

```
  test('renders default state correctly', () => {
    const tree = renderer.create(<Provider store={mockStore()}>
      <JobForm />
    </Provider>).toJSON()

    expect(tree).toMatchSnapshot()
  })
```

gives `Expected closing tag to match indentation of opening  react/jsx-closing-tag-location` error. If you fix that, you get `Expected indentation of 4 space characters but found 33  react/jsx-indent` error.

**Comment by ljharb** [comment]
```jsx
  test('renders default state correctly', () => {
    const tree = renderer.create(
      <Provider store={mockStore()}>
        <JobForm />
      </Provider>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
```

**Comment by mrchief** [comment]
2 issues with that:

`eslint --fix` formats that to

```
    const tree = renderer.create(<Provider store={mockStore()}>
      <JobForm />
                                 </Provider>).toJSON()
```

which then gives the `react/jsx-indent` error.

If you run `eslint` without the autofix. you get `error  Unexpected newline after '('  function-paren-newline` error.

Been struggling with this for past few hours. :(

**Comment by ljharb** [comment]
@mrchief then try:
```jsx
 test('renders default state correctly', () => {
    const tree = renderer.create((
      <Provider store={mockStore()}>
        <JobForm />
      </Provider>
    )).toJSON()

    expect(tree).toMatchSnapshot()
  })
```

**Comment by mrchief** [comment]
@ljharb That works. I'm not sure if I like the extraneous parens. 

And the fact that I've to go rewrite all my existing code - which I think I'll have to rollback in some future update that'll complain about the extra parens.

**Comment by rbalicki2** [comment]
I, too, would like `line-aligned`. I don't think it's a niche requirement, TBH.

Adding extra line breaks so that the opening tag and the closing tag can be in the same column adds a lot visual bloat that doesn't need to be there.

**Comment by andy-vacasa** [comment]
+1

**Comment by ljharb** [comment]
@andy-vacasa github's had a reaction emoji feature for over a year now; please use that instead of useless "+1" comments. I'll now delete your comment and my reply.

**Comment by ezzatron** [comment]
@ljharb Can you expand on the reasons why you recommend wrapping multiline JSX in parens? The `react/jsx-wrap-multilines` rule [docs](https://github.com/yannickcr/eslint-plugin-react/blob/4a719e8/docs/rules/jsx-wrap-multilines.md) are pretty vague/subjective:

> Wrapping multiline JSX in parentheses can improve readability and/or convenience.

I did a bit of my own research, and found some potential reasons that have to do with automatic semicolon insertion for return statements:

- [Why use parenthesis on JavaScript return statements?](http://jamesknelson.com/javascript-return-parenthesis/)
- [This comment on a prettier/prettier issue](https://github.com/prettier/prettier/issues/1009#issuecomment-286833002)

But if that's the reason for the recommendation, wouldn't it be better / more comprehensive to simply lint for unreachable code? Happy to have my opinion changed btw, I just want to understand better.

**Comment by ljharb** [comment]
@ezzatron a single JSX element (containing all its children) is a single method. All multiline constructs should be wrapped by parens, brackets, or braces so that it's explicitly clear what the single value is. Certainly this helps with ASI hazards, but it also visually helps group a JSX element without needing to look solely for the closing jsx tag.

**Comment by ljharb** [comment]
@Chudesnov it's still relevant because things inside `{ }`, in a jsx context, follow JS rules, whereas everything else in a jsx context follows jsx rules. jsx hides the function calls; it *can not* hide the cognitive overhead of needing to know about the contexts and when to switch between them.

**Comment by golopot** [comment]
My opinion is that this rule should stop checking the column location of the closing tag, and leave the indent dealt by `indent` or `jsx-indent`. And so that this rule only check that closing tag is in a new line.

**Comment by rojobuffalo** [comment]
I'm supportive of the OP request. I really like not wrapping jsx in parens and satisfying line-aligned.

```
return <div>
</div>;
```

**Comment by qfox** [comment]
Heya. Any activity here?

**Comment by pascalpp** [comment]
Here's an example that might clarify why a line-aligned option would be useful:

```
function SupportLink() {
  return (
    <p className="support-link">
      Tip: <Link
        className="use-template"
        label="See tips to get the most out of campaigns"
        href="https://example.com"
        external>
          Get the most out of campaigns.
      </Link>
    </p>
  );
}
```

The tag is preceded by a word, and the whitespace between that word and the tag is critical to how it is rendered.

Since this is fighting with `indent`, I like @golopot 's suggestion to just ensure that the closing tag is either:
- on the same line as the opening tag
- or on its own line

**Comment by ljharb** [comment]
@pascalpp here's how i'd write that (and how the airbnb styleguide would mandate you write it):
```jsx
function SupportLink() {
  return (
    <p className="support-link">
      Tip: {(
        <Link
          className="use-template"
          label="See tips to get the most out of campaigns"
          href="https://example.com"
          external>
            Get the most out of campaigns.
        </Link>
      )}
    </p>
  );
}
```

**Comment by rojobuffalo** [comment]
^ respectfully, i do not like that syntax.  imo the extra characters and nesting are extraneous. the syntax of an element beginning and ending with `<` and `>` is all the demarcation i want. it's a difference of a boundary being one character vs. 4 characters: `<` vs. `{`, `(`, `\n` `^I`

**Comment by ljharb** [comment]
Then put a newline after “<“ so that the “Link” can start on a new line?

The length of the boundary isn’t important; what’s important is the clean demarcation of contexts.

Either way, since the spacing here is actually important, using bare text is actually hiding the intent:

```jsx
function SupportLink() {
  return (
    <p className="support-link">
      {'Tip: '}
      <Link
        className="use-template"
        label="See tips to get the most out of campaigns"
        href="https://example.com"
        external
      >
          Get the most out of campaigns.
      </Link>
    </p>
  );
}
```

**Comment by rojobuffalo** [comment]
in these examples, any characters besides the element tag are extraneous in my opinion, i.e. a newline is extraneous. the `<` character (as in `<Link>`) is present no matter what so that's irreducible.
if my preference for a code style is to only use the irreducible parts and not decorate or wrap with optional syntax or demarcations, i feel like that should be supported by an eslint option.
re: a whitespace after "Tip: ", it is obvious to me that it's a space character that will be rendered. i don't want/need more characters to point that out

**Comment by ljharb** [comment]
It doesn't matter what's obvious to you - the goal is to make the code readable to *someone else*.
